### PR TITLE
[`MobileSam`] Adds MobileSAM to transformers

### PR DESCRIPTION
**Pull request by younesbelkada**
# What does this PR do?

as discussed offline cc @merveenoyan @NielsRogge 
This PR adds MobileSam to the library. MobileSam uses the same archtiecture as SAM, with the SAM image encoder being swapped to TinyViT. Therefore I decided to create a new modeling file for it, as porting TinyViT required a bit of work

Draft for now!

### CODE REVIEW

**Comment by nidhishgajjar** [comment]
> **Orb Code Review** (powered by GLM 5.1 on [Orb Cloud](https://orbcloud.dev))

Thanks for adding MobileSAM to transformers! I've reviewed the modeling code, configuration, and tests. There are several critical issues that need to be addressed before this can be merged.

---

### 🐛 Must-Fix: Broken `MobileSamVisionAttention.forward()`

The `MobileSamVisionAttention` module has **undefined attributes** that will cause `AttributeError` at runtime:

**`self.d` never defined** (line ~1069):
```python
q, k, v = qkv.view(B, N, self.num_attention_heads, -1).split(
    [self.hidden_dim, self.hidden_dim, self.d], dim=3
)
```
`self.d` is never set in `__init__`. This should be `head_dim = self.hidden_dim // self.num_attention_heads`.

**`self.attention_bias_idxs` and `self.ab` never defined** (line ~1078):
```python
(self.attention_biases[:, self.attention_bias_idxs]
 if self.training else self.ab)
```
Neither attribute exists. The original TinyViT implementation pre-computes index tensors for relative position biases — those need to be registered here (e.g., via `register_buffer`).

**`self.dh` never defined** (line ~1082):
```python
x = (attn @ v).transpose(1, 2).reshape(B, N, self.dh)
```
Should likely be `self.hidden_dim`.

The `self.dropout` attribute is set but never applied in the forward pass.

---

### 🐛 Must-Fix: Bad find-replace "downsample" → "downmobilesample"

The SAM→MobileSam find-replace accidentally renamed `downsample` to `downmobilesample` throughout the mask decoder code:

```python
# In config:
attention_downmobilesample_rate=2  # should be attention_downsample_rate

# In modeling:
downmobilesample_rate=None  # should be downsample_rate
```

This appears in `MobileSamMaskDecoderConfig`, `MobileSamAttention`, `MobileSamTwoWayAttentionBlock`, `MobileSamTwoWayTransformer`, and the test file. All instances should be corrected to use `downsample`.

---

### 🐛 Must-Fix: `MobileSamVisionEncoder.forward()` doesn't collect hidden states or attentions

The encoder initializes `all_hidden_states = ()` and `all_self_attentions = ()` but never appends to them in the layer loop:

```python
for layer in self.layers:
    hidden_states = layer(hidden_states)
    # missing: all_hidden_states += (hidden_states,)
    # missing: attention collection from layers
```

This means `output_hidden_states=True` and `output_attentions=True` always return empty tuples, which breaks the corresponding tests.

---

### ⚠️ Cleanup: Unused classifier head in `MobileSamVisionEncoder`

```python
self.layer_norm_head = nn.LayerNorm(config.embed_dims[-1])
self.head = nn.Linear(config.embed_dims[-1], config.num_classes)
```

These appear to be leftover from the original TinyViT classification model and are never used in the forward pass. They add unnecessary parameters. `config.num_classes` (default 1000) also doesn't belong in a segmentation model config.

---

### 📝 Minor Notes

- **Typo**: Docstring says `*optionalk*` instead of `*optional*` in `MobileSamTwoWayAttentionBlock.__init__`
- **Extra blank line** in `MobileSamVisionConv.forward()` (between batch_norm and return)
- **Consistency**: The codebase uses "MOBILESAM" (all caps) in many places but the model name is "MobileSAM" — worth aligning these

---

### Summary

The PR adds a comprehensive MobileSAM implementation with config, modeling, tests, and conversion scripts. However, the `MobileSamVisionAttention` class has critical bugs (undefined `self.d`, `self.dh`, `self.ab`, `self.attention_bias_idxs`) that will cause runtime errors, and the "downmobilesample" naming issue suggests the SAM→MobileSam copy was done via find-replace without sufficient review. These need to be fixed before the model can work correctly.

### Assessment: request-changes

_This review was created by an AI assistant (OpenHands) on behalf of Orb Cloud._
